packages/cli/src/models/compiler/Compiler.ts

```diff
@@ -1,35 +1,67 @@
+import findUp from 'find-up';
+import ZosPackageFile from '../files/ZosPackageFile';
 
 const log = new Logger('Compiler');
-
 const state = { alreadyCompiled: false };
 
-const Compiler = {
-  async call(force: boolean = false): Promise<{ stdout: string, stderr: string }> {
-    if (force || !state.alreadyCompiled) {
-      log.info('Compiling contracts with Truffle...');
-      let truffleBin = `${process.cwd()}/node_modules/.bin/truffle`;
-      if (!FileSystem.exists(truffleBin)) truffleBin = 'truffle'; // Attempt to load global truffle if local was not found
-
-      return new Promise((resolve, reject) => {
-        const args: object = { shell: true };
-
-        execFile(truffleBin, ['compile', '--all'], args, (error: ExecException, stdout, stderr) => {
-
-          if (error) {
-            if (error.code === 127) console.error('Could not find truffle executable. Please install it by running: npm install truffle');
-            reject(error);
-          } else {
-            state.alreadyCompiled = true;
-            resolve({ stdout, stderr });
-          }
-          if (stdout) console.log(stdout);
-          if (stderr) console.error(stderr);
-        });
-      });
-    }
+export async function compile(compilerOptions?: ProjectCompilerOptions, packageFile = new ZosPackageFile(), force: boolean = false): Promise<void> {
+  if (!force && state.alreadyCompiled) return;
+
+  // Merge config file compiler options with those set explicitly
+  compilerOptions = { ...packageFile.compilerOptions, ...compilerOptions };
+
+  // Validate compiler manager setting
+  const { manager } = compilerOptions;
+  if (manager && manager !== 'truffle' && manager !== 'zos') {
+    throw new Error(`Unknown compiler manager '${manager}' (valid values are 'zos' or 'truffle')`);
   }
-};
 
-export default Compiler;
+  // We use truffle if set explicitly, or if nothing was set but there is a truffle.js file
+  const useTruffle = manager === 'truffle'
+    || (!manager && Truffle.isTruffleProject());
+
+  // Compile! We use the exports syntax so we can stub them out during tests (nasty, but works!)
+  const compilePromise = useTruffle
+    ? exports.compileWithTruffle()
+    : exports.compileWithSolc(compilerOptions);
+  await compilePromise;
+
+  // If compiled successfully, write back compiler settings to zos.json to persist them
+  packageFile.setCompilerOptions({ ... compilerOptions, manager: useTruffle ? 'truffle' : 'zos' });
+  packageFile.write();
+
+  state.alreadyCompiled = true;
+}
+
+export async function compileWithSolc(compilerOptions?: ProjectCompilerOptions): Promise<void> {
+  await compileProject(compilerOptions);
+}
+
+export async function compileWithTruffle(): Promise<void> {
+  log.info('Compiling contracts with Truffle...');
+  // Attempt to load global truffle if local was not found
+  const truffleBin: string = findUp.sync('node_modules/.bin/truffle') || 'truffle';
```

Comment:
I usually prefer sync versions if available, instead of awaiting on asyncs. I have absolutely zero evidence of it, but I'd say that the OS is better at awaiting than node.

Reply:
You mean that sync is faster than async because of all the context switching due message queue? Probably. But blocking the only node.js thread will clog message queue and will cause other tasks relying on it to suffer.